### PR TITLE
Fix #3527

### LINKED ISSUE #3527: Update baseline script to support python 3
The baseline script https://github.com/zaproxy/zaproxy/blob/develop/build/docker/zap-baseline.py currently only supports python 2.
It would be good for this to be python 3 compatible as well.
Hopefully that wont be too hard for a pythonista?

### PR DESCRIPTION
**Pull request by AnamikaD**
@thc202 @psiinon I made the required changes. I am using python and python3, but I get the same error as "no module zap"

### CODE REVIEW

**Comment by AnamikaD** [comment]
For issue #119 
I copied the code mentioned in the comments. I get this error
````
ubuntu@ubuntu:~/zaproxy/src$ ./zap.sh 
Found Java version 1.8.0_111
Available memory: 1992 MB
Setting jvm heap size: -Xmx498m
Error: Unable to access jarfile /home/ubuntu/zaproxy/src/zap-dev.jar

**Comment by kingthorin** [comment]
@AnamikaD thanks for the contributions. Need you to address a few things though.

1. A pull request should be for a single issue.
2. New extensions should be added to zaproxy/zap-extensions (alpha). [Also note the patch you grabbed out of the issue is very old and likely does not account for newer functionality/framework.]
3. Formatting and white space changes should happen separately from code changes. (https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#formatting)
4. For the python stuff you might want to maintain compatibility with both python 2 and 3, something like: https://stackoverflow.com/a/17510727/7718222 (If use of a third part library is really necessary then the commit/PR message should outline why, and the docker build files should be updated to include installation of that library.)

**Comment by AnamikaD** [comment]
@kingthorin thank you for the feedback!! I would make the changes and update the PR.

**Comment by thc202** [comment]
@AnamikaD could you provide the output of the following commands:
```
pip --version
pip list | grep zap
```
and 
```
pip3 --version
pip3 list | grep zap
```

@kingthorin regarding 4., `six` is already a dependency of ZAP API Python client.

**Comment by thc202** [comment]
@AnamikaD regarding https://github.com/zaproxy/zaproxy/pull/3617#issuecomment-306263370, to run ZAP from the command line you need to build it first [1] and then run the result (under build/zap).

[1] https://github.com/zaproxy/zaproxy/wiki/Building#ant-no-ide

**Comment by kingthorin** [comment]
> @kingthorin regarding 4., six is already a dependency of ZAP API Python client.
Ok fair enough.

**Comment by AnamikaD** [comment]
@thc202  Thank you for the reply! For pip I get error. I think that it is not installed
```
ubuntu@ubuntu:~$ pip3 --version
pip 1.5.6 from /usr/lib/python3/dist-packages (python 3.5)
ubuntu@ubuntu:~$ pip3 list | grep zap
python-owasp-zap-v2.4 (0.0.10)

**Comment by AnamikaD** [comment]
pip output
```
ubuntu@ubuntu:~$ pip --version
Traceback (most recent call last):
  File "/usr/bin/pip", line 9, in <module>
    load_entry_point('pip==1.5.6', 'console_scripts', 'pip')()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 558, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2682, in load_entry_point
    return ep.load()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2355, in load
    return self.resolve()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2361, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/lib/python2.7/dist-packages/pip/__init__.py", line 74, in <module>
    from pip.vcs import git, mercurial, subversion, bazaar  # noqa
  File "/usr/lib/python2.7/dist-packages/pip/vcs/mercurial.py", line 9, in <module>
    from pip.download import path_to_url
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 25, in <module>
    from requests.compat import IncompleteRead
ImportError: cannot import name IncompleteRead

**Comment by AnamikaD** [comment]
@kingthorin  I am unable to find the path "zaproxy/zap-extensions".

**Comment by kingthorin** [comment]
No problem, you probably need to fork and setup the extensions repo:
https://github.com/zaproxy/zaproxy/wiki/Contributing-Changes#fork-the-relevant-projects-to-your-own-github-user

**Comment by AnamikaD** [comment]
@kingthorin I had forked this "https://github.com/zaproxy/zaproxy" repository. The extension is here
"zaproxy/src/org/zaproxy/zap/extension/"

**Comment by thc202** [comment]
It's another repo:

**Comment by AnamikaD** [comment]
@thc202 Thanks!!

**Comment by AnamikaD** [comment]
@thc202 
1) In zaproxy/zaproxy, after a successful build, I get the same error.
2) After the changes in  zaproxy/zap-extension, How do I run it? I have build it.

**Comment by kingthorin** [comment]
For no. 2. new extensions should be in the alpha branch. https://github.com/zaproxy/zap-extensions/blob/alpha/build/build.xml should be updated with the appropriate details (for example: https://github.com/zaproxy/zap-extensions/blob/alpha/build/build.xml#L547-L554). You'd then right click on the new ant build task and "Run as ant build". Then launch ZAP and the addon should be loaded, if it isn't then hit ctrl+L and navigate to the build location (ant displays this in the eclipse console while running the task), select the addon, and open (load) it...

**Comment by thc202** [comment]
For 1. are you running the zap.sh in the build/zap dir?

**Comment by thc202** [comment]
@AnamikaD regarding `pip`, how was that installed, with package manager? or "manually"?

**Comment by AnamikaD** [comment]
@thc202 I am not sure about the installation but would remove pip and install again.
Yes, running under  build/zap and it is working now! Thank you

**Comment by AnamikaD** [comment]
@kingthorin I don't have eclipse right now! I build it using ant. There is no zap.sh file in zap-extension

**Comment by kingthorin** [comment]
Ok no problem, if you've built the extension using ant. Run ZAP (as you normally would) and load the extension and outlined earlier.

**Comment by AnamikaD** [comment]
@thc202 I have reinstalled pip : sudo apt-get install python-pip
I get the same error message. Any other solution??

**Comment by AnamikaD** [comment]
@kingthorin Please review the code for issue #119  [here](https://github.com/zaproxy/zap-extensions/pull/932).

**Comment by thc202** [comment]
@AnamikaD regarding the pip error it seems to be <https://bugs.launchpad.net/ubuntu/+source/python-pip/+bug/1306991>. Might worth try installing it manually: https://pip.pypa.io/en/latest/installing/#installing-with-get-pip-py (after removing the one installed with package manager).

**Comment by AnamikaD** [comment]
@thc202 It was helpful! can you review the code??

File: build/docker/zap-baseline.py [hunk 1]
```diff
@@ -354,13 +357,13 @@ def main(argv):
...
-                if (not alert_dict.has_key(plugin_id)):
+                if (plugin_id not in alert_dict):
                     pass_dict[plugin_id] = rule.get('name')
 
             if min_level == levels.index("PASS") and detailed_output:
-                for key, rule in sorted(pass_dict.iteritems()):
+                for key, rule in sorted(pass_dict.items()):
                     print('PASS: ' + rule + ' [' + key + ']')
-
+
```
Comment (thc202) [inline build/docker/zap-baseline.py]: This indentation change should be reverted.

**Comment by AnamikaD** [comment]
@thc202 Thank you for the review!! I would check it. Is it working for python 2?

**Comment by thc202** [comment]
Yes, the current code is working fine with Python 2.

**Comment by AnamikaD** [comment]
@thc202  I have fixed those errors but unable to run it. It shows error related to zap module.

**Comment by thc202** [comment]
OK, did you try installing `pip` with `get-pip.py`?

Regarding Python 3, there's some changes required in `zap_common.py`, getting:
```
ERROR <class 'AttributeError'>
2017-06-15 10:56:56,812 Unexpected error: <class 'AttributeError'>
Traceback (most recent call last):
  File "build/docker/zap-baseline.py", line 371, in main
    inc_ignore_rules, True, detailed_output, {})
  File "/zaproxy/build/docker/zap_common.py", line 94, in print_rules
    for key, alert_list in sorted(alert_dict.iteritems()):
AttributeError: 'dict' object has no attribute 'iteritems'
```

**Comment by AnamikaD** [comment]
pip --version : pip 9.0.1
I get this output when run through python3
```
ubuntu@ubuntu:~/zaproxy/build/docker$ python3 zap-baseline.py 
Traceback (most recent call last):
  File "zap-baseline.py", line 56, in <module>
    from zapv2 import ZAPv2
  File "/home/ubuntu/.local/lib/python3.5/site-packages/zapv2/__init__.py", line 25, in <module>
    import requests
  File "/home/ubuntu/.local/lib/python3.5/site-packages/requests/__init__.py", line 78, in <module>
    from urllib3.contrib import pyopenssl
  File "/home/ubuntu/.local/lib/python3.5/site-packages/urllib3/contrib/pyopenssl.py", line 46, in <module>
    import OpenSSL.SSL
  File "/usr/local/lib/python3.5/dist-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import rand, crypto, SSL
  File "/usr/local/lib/python3.5/dist-packages/OpenSSL/SSL.py", line 112, in <module>
    if _lib.Cryptography_HAS_SSL_ST:
AttributeError: module 'lib' has no attribute 'Cryptography_HAS_SSL_ST'

**Comment by thc202** [comment]
Could you try update the cryptography package:
`pip3 install -U cryptography`

There are still some usage of `has_key` in `zap_common.py` that's causing errors.

**Comment by AnamikaD** [comment]
Python3 does not support "has_key" ( "in" is used)
I get this error for cryptography. 
```
Command /usr/bin/python3 -c "import setuptools, tokenize;__file__='/tmp/pip-build-3uuxjtaj/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-w_cmhysy-record/install-record.txt --single-version-externally-managed --compile --user failed with error code 1 in /tmp/pip-build-3uuxjtaj/cryptography
Storing debug log for failure in /tmp/tmp_n90w0zl

File: build/docker/zap_common.py [hunk 2]
```diff
@@ -107,19 +107,19 @@ def print_rules(alert_dict, level, config_dict, config_msg, min_level, levels, i
...
 
 
 def inc_info_rules(config_dict, key, inc_extra):
-    return (config_dict.has_key(key) and config_dict[key] == 'INFO') or (not config_dict.has_key(key) and inc_extra)
+    return (key in config_dict) and config_dict[key] == 'INFO') or ((key not in config_dict) and inc_extra)
 
 
 def inc_warn_rules(config_dict, key, inc_extra):
-    return (config_dict.has_key(key) and config_dict[key] == 'WARN') or (not config_dict.has_key(key) and inc_extra)
+    return (key in config_dict) and config_dict[key] == 'WARN') or ((key in config_dict) and inc_extra)
```
Comment (thc202) [inline build/docker/zap_common.py]: In the right change it's missing the "not".

**Comment by thc202** [comment]
:( what does it say in the log?

File: build/docker/zap_common.py [hunk 3]
```diff
@@ -107,19 +107,19 @@ def print_rules(alert_dict, level, config_dict, config_msg, min_level, levels, i
 
 
 def inc_ignore_rules(config_dict, key, inc_extra):
-    return config_dict.has_key(key) and config_dict[key] == 'IGNORE'
+    return (key in config_dict) and config_dict[key] == 'IGNORE'
 
 
 def inc_info_rules(config_dict, key, inc_extra):
-    return (config_dict.has_key(key) and config_dict[key] == 'INFO') or (not config_dict.has_key(key) and inc_extra)
+    return (key in config_dict) and config_dict[key] == 'INFO') or ((key not in config_dict) and inc_extra)
```
Comment (thc202) [inline build/docker/zap_common.py]: It's missing one opening parenthesis in the left expression, or remove the closing parenthesis before "and"?

File: build/docker/zap_common.py (hunk 2 above)
Comment (thc202) [inline build/docker/zap_common.py]: Same comment regarding the missing parenthesis.

File: build/docker/zap_common.py (hunk 3 above)
Comment (AnamikaD) [inline build/docker/zap_common.py]: Sorry I forgot to put a build.

**Comment by thc202** [comment]
There's just two things left (AFAICT):
```
zap-baseline.py:409: DeprecationWarning: __getitem__ not supported for exception classes in 3.x; use args attribute
errno, strerror = e
```
and remove the whitespaces added to an empty line.

**Comment by AnamikaD** [comment]
@thc202 I checked for whitespaces but did not find it

**Comment by thc202** [comment]
It's in the file `zap-baseline.py`: https://github.com/zaproxy/zaproxy/pull/3617#discussion_r121484730

**Comment by AnamikaD** [comment]
@thc202 Thanks alot!!

**Comment by thc202** [comment]
Thank you! Seems to be running fine with Python 2 and 3.

Would you mind fix up the commits?

**Comment by AnamikaD** [comment]
@thc202 Thank you so much! how do I fix the commits?

**Comment by kingthorin** [comment]
Do an interactive rebase and fix-up or squash into a single commit.

- https://www.instructables.com/id/How-to-Rebase-Interactive-With-Eclipse-eGit/
- https://www.kernel.org/pub/software/scm/git/docs/git-rebase.html#_interactive_mode
- https://wiki.eclipse.org/EGit/User_Guide#Interactive_Rebase

**Comment by AnamikaD** [comment]
@kingthorin Thanks! I will do that now.

**Comment by thc202** [comment]
Thank you!

**Comment by AnamikaD** [comment]
@thc202 thank you!!

**Pull request merged**